File: src/main/java/io/shiftleft/fuzzyc2cpg/output/inmemory/OutputModule.java
package io.shiftleft.fuzzyc2cpg.output.inmemory;

import io.shiftleft.codepropertygraph.Cpg;
import io.shiftleft.codepropertygraph.cpgloading.ProtoCpgLoader;
import io.shiftleft.fuzzyc2cpg.output.CpgOutputModule;
import io.shiftleft.passes.DiffGraph;
import io.shiftleft.passes.KeyPool;
import java.io.IOException;
import overflowdb.OdbConfig;
import io.shiftleft.proto.cpg.Cpg.CpgStruct;

import java.util.LinkedList;
import java.util.List;

public class OutputModule implements CpgOutputModule {

  private LinkedList<CpgStruct.Builder> cpgBuilders;
  private Cpg cpg;

  protected OutputModule() {
    this.cpgBuilders = new LinkedList<>();
  }

  public Cpg getInternalGraph() {
    return cpg;
  }

  @Override
  public void persistCpg(CpgStruct.Builder cpg, String identifier) {
    synchronized (cpgBuilders) {
      cpgBuilders.add(cpg);
    }
  }

  @Override
  public void persistCpg(DiffGraph diffGraph, KeyPool keyPool, String identifier) throws IOException {
    // TODO
  }

  public void persist() {
    CpgStruct.Builder mergedBuilder = CpgStruct.newBuilder();

    cpgBuilders.forEach(builder -> {
      mergedBuilder.mergeFrom(builder.build());
    });

    List<CpgStruct> list = new LinkedList<>();
    list.add(mergedBuilder.build());
    cpg = ProtoCpgLoader.loadFromListOfProtos(list, OdbConfig.withoutOverflow());
  }

}

File: src/main/scala/io/shiftleft/fuzzyc2cpg/output/inmemory/OutputModule.scala
package io.shiftleft.fuzzyc2cpg.output.inmemory

import io.shiftleft.codepropertygraph.Cpg
import io.shiftleft.codepropertygraph.cpgloading.ProtoToCpg
import io.shiftleft.fuzzyc2cpg.output.CpgOutputModule
import io.shiftleft.passes.{DiffGraph, KeyPool}
import java.io.IOException

import io.shiftleft.fuzzyc2cpg.output.overflowdb.DiffGraphAndKeyPool
import overflowdb.OdbConfig
import io.shiftleft.proto.cpg.Cpg.CpgStruct
import scala.collection.mutable.ListBuffer

class OutputModule extends CpgOutputModule {

  private val cpgBuilders =
    ListBuffer[Either[CpgStruct.Builder, DiffGraphAndKeyPool]]()
  private val odbConfig = OdbConfig.withoutOverflow()
  private val protoToCpg = new ProtoToCpg(odbConfig)

  def getInternalGraph: Cpg = new Cpg(protoToCpg.graph)

  override def persistCpg(builder: CpgStruct.Builder): Unit = {
    cpgBuilders synchronized {
      cpgBuilders.addOne(Left(builder))
    }
  }

  @throws[IOException]
  override def persistCpg(diffGraph: DiffGraph, keyPool: KeyPool): Unit = {
    cpgBuilders synchronized {
      cpgBuilders.addOne(Right(DiffGraphAndKeyPool(diffGraph, keyPool)))
    }
  }

  def persist(): Unit = {
    cpgBuilders.foreach {
      case Left(b) =>
        protoToCpg.addNodes(b.getNodeList)
        protoToCpg.addEdges(b.getEdgeList)
      case Right(DiffGraphAndKeyPool(diffGraph, keyPool)) =>
        DiffGraph.Applier.applyDiff(diffGraph, protoToCpg.graph, undoable = false, Some(keyPool))
    }
  }
}
